### ISSUE TITLE
Prevent starvation of large jobs

### ISSUE THREAD

**Issue by ahg-g** [kubernetes-sigs/kueue#534]
**What would you like to be added**:

Workloads with the same priority don't preempt each other within a ClusterQueue, one consequence is that large size jobs may be starved if users submit a stream of smaller jobs from the same priority class. 

To address this case, we could add a preemption strategy that allows older jobs to preempt newer ones from the same priority class.

**Why is this needed**:

To prevent starvation of large jobs via preemption. 

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

**Comment by alculquicondor** [comment]
The field `TriggerAfterWorkloadWaitingSeconds` in [1] was a proposal along these lines. Any extra signal we can incorporate in a new design?

[1] https://github.com/kubernetes-sigs/kueue/tree/main/keps/83-workload-preemption#extra-knobs-in-clusterqueue-preemption-policy

**Comment by ahg-g** [comment]
Yes, so smaller jobs that were submitted after the big job, but got a chance to start before the big one because they happen to fit, should be the primary candidates for preemption. 

Imagine this example (all jobs from the same priority class):
1) quota is 100
2) job1 and job2 each consumes 30; available quota is 40
3) job3 submitted and requests 50, it doesn't fit and will wait
4) job4 submitted and requests 30, it fits and will start; available quota is 10
5) job1 finishes, available quota is 40, so job3 still doesn't fit, but we can preempt job4 to fit it; job4 is a candidate because it was submitted after job3.

**Comment by alculquicondor** [comment]
Gotcha, that's the creation timestamps for each Jobs is a very clear signal.

**Comment by KunWuLuan** [comment]
Maybe we can add a property in clusterQueue to allow workload preempting workloads with same priority after some time?

**Comment by trasc** [comment]
A middle ground between this and `TriggerAfterWorkloadWaitingSeconds` could be someting like:
```go
	// PreemptSmallerIfNewerThen if set, smaller jobs having a scheduling timestamp
	// bigger than the current workload + PreemptSmallerIfNewerThen seconds will be
	// considered for preemption in order to avoid starvation of large jobs.
	// This is only applicable within the same cluster queue when WithinClusterQueue
	// is set to LowerPriority.
	//
	// +optional
	PreemptSmallerIfNewerThen *int64

```

**Comment by alculquicondor** [comment]
This should be a Clusterqueue API field.
I like the flexibility of the proposal above. However, the wording needs to include that this affects the same priority. And maybe that this is specific to ClusterQueue preemption, and not cohort reclaim? does it have to be?

Also, do we expect more tuning options? If so, it could be something like:

```yaml
preemption:
  withinClusterQueue: LowerThanOrEqualPriority
  samePriority:
     ifNewerThanSeconds: 100
```

Although I'm not convinced about the wording

**Comment by alculquicondor** [comment]
Let's step back.
@KunWuLuan, do you have a use case for configuring the time threshold?
We could start simple and just add the policy for LowetThanOrEqualPriority

**Comment by trasc** [comment]
I have opened #710, but I'd hold the merge and use workload.GetSchedulingTimestamp after #689 is merged.

**Comment by KunWuLuan** [comment]
> Let's step back. @KunWuLuan, do you have a use case for configuring the time threshold? We could start simple and just add the policy for LowetThanOrEqualPriority

Hi, Aldo. I think `ifNewerThanSeconds ` is enougth. Thanks. 👍

**Comment by alculquicondor** [comment]
My question was whether we could start without `ifNewerThanSeconds`.
Can you clarify why do you think this is important?
Could there be other alternatives that apply to any kind of preemption? Like `TriggerAfterWorkloadWaitingSeconds` in [1]
Or something like `IfAdmittedLessThanSeconds`.

[1] https://github.com/kubernetes-sigs/kueue/tree/main/keps/83-workload-preemption#extra-knobs-in-clusterqueue-preemption-policy
